lib/spack/spack/schema/concretizer.py

```diff
@@ -9,13 +9,40 @@
 """
 from typing import Any, Dict
 
+LIST_OF_SPECS = {"type": "array", "items": {"type": "string"}}
+
 properties: Dict[str, Any] = {
     "concretizer": {
         "type": "object",
         "additionalProperties": False,
         "properties": {
             "reuse": {
-                "oneOf": [{"type": "boolean"}, {"type": "string", "enum": ["dependencies"]}]
+                "oneOf": [
+                    {"type": "boolean"},
+                    {"type": "string", "enum": ["dependencies"]},
+                    {
+                        "type": "object",
+                        "properties": {
+                            "roots": {"type": "boolean"},
+                            "include": LIST_OF_SPECS,
+                            "exclude": LIST_OF_SPECS,
+                            "from": {
+                                "type": "array",
+                                "items": {
+                                    "type": "object",
+                                    "properties": {
+                                        "type": {
+                                            "type": "string",
+                                            "enum": ["local", "mirror", "external"],
```

Comment:
I know I'm late to the review, but how much of a lift would it be to add `upstreams` to this list? Is the intent that those would be coupled under `local`?  I would personally like to see them differentiated if possible.

Reply:
> I would personally like to see them differentiated if possible.

We can have additional PRs later, to add more control - but I'll keep this one as is :slightly_smiling_face: What would be the use case for distinguishing upstreams from local?

Reply:
Primarily to have more control. Say I suspect there might be an issue with the upstream, or I wanted to test phasing one out. Follow on is fine by me.

Reply:
See #42999 -- a symmetric config option here makes a lot of sense. Upstreams have names and paths that they can be identified by (kind of like mirrors)